src/util/safe_print.cpp

```diff
@@ -25,20 +28,20 @@
 namespace CVC4 {
 
 template <>
-void safe_print(int fd, const char* msg) {
-  ssize_t ret = write(fd, msg, safe_strlen(msg));
-  (void)ret;
-}
-
-template <>
-void safe_print(int fd, const std::string* msg) {
-  ssize_t ret = write(fd, msg->c_str(), msg->length());
-  (void)ret;
+void safe_print(int fd, const std::string& msg) {
+  // Print characters one by one instead of using
+  // string::data()/string::c_str() to avoid allocations (pre-c++11)
+  for (size_t i = 0; i < msg.length(); i++) {
+    if(write(fd, &(msg[i]), 1) != 1) {
+      abort();
+    }
+  }
 }
 
 template <>
-void safe_print(int fd, int64_t i) {
+void safe_print(int fd, const int64_t& _i) {
```

Comment:
It is to match the declaration of `safe_print`: `template <typename T>
 void CVC4_PUBLIC safe_print(int fd, const T& msg)`. I agree that it is somewhat strange to pass primitive types by reference. Do you have a better idea?
There are two alternatives that I can think of right now:
- We could basically do what we had before and only pass certain types by reference but then we need to add special cases in `statistics_registry.h`, which is also not great
- Not use templates at all (and just define multiple versions of `safe_print`) but then it is much less clear (IMHO) to which version the `safe_print` calls in `statistics_registry.h` get dispatched and it might lead to some weirdnesses with implicit casting (not sure)